diff --git kernel/fiasco/src/kern/arm/thread-arm.cpp kernel/fiasco/src/kern/arm/thread-arm.cpp
index 8dc4cbf..76d3a9f 100644
--- kernel/fiasco/src/kern/arm/thread-arm.cpp
+++ kernel/fiasco/src/kern/arm/thread-arm.cpp
@@ -315,6 +315,10 @@ extern "C" {
             if (handle_copro_fault[copro](opcode, ts))
               return;
           }
+
+        /* check for ARM default GDB breakpoint */
+        if (!(ts->psr & Proc::Status_thumb) && opcode == 0xe7ffdefe)
+          ts->pc -= 4;
       }
 
 undef_insn:
