diff --git kernel/fiasco/src/kern/arm/thread-arm.cpp kernel/fiasco/src/kern/arm/thread-arm.cpp
index 76d3a9f..a781088 100644
--- kernel/fiasco/src/kern/arm/thread-arm.cpp
+++ kernel/fiasco/src/kern/arm/thread-arm.cpp
@@ -418,7 +418,7 @@ Thread::user_ip() const
 IMPLEMENT inline
 Mword
 Thread::user_flags() const
-{ return 0; }
+{ return state() & Thread_ready; }
 
 IMPLEMENT inline NEEDS[Thread::exception_triggered]
 void
diff --git kernel/fiasco/src/kern/ia32/thread-ia32.cpp kernel/fiasco/src/kern/ia32/thread-ia32.cpp
index d6fb000..f494a4d 100644
--- kernel/fiasco/src/kern/ia32/thread-ia32.cpp
+++ kernel/fiasco/src/kern/ia32/thread-ia32.cpp
@@ -92,7 +92,7 @@ Thread::user_ip() const
 IMPLEMENT inline
 Mword
 Thread::user_flags() const
-{ return regs()->flags(); }
+{ return state() & Thread_ready; }
 
 /** Check if the pagefault occured at a special place: At some places in the
     kernel we want to ensure that a specific address is mapped. The regular
