diff --git kernel/fiasco/src/kern/arm/bsp/imx/Kconfig kernel/fiasco/src/kern/arm/bsp/imx/Kconfig
index 5ec8388..1312b32 100644
--- kernel/fiasco/src/kern/arm/bsp/imx/Kconfig
+++ kernel/fiasco/src/kern/arm/bsp/imx/Kconfig
@@ -43,6 +43,13 @@ config PF_IMX_6
         help
           Choose for i.MX6 platform.
 
+config PF_IMX_53
+	bool "i.MX53"
+	depends on PF_IMX
+	select CAN_ARM_CPU_CORTEX_A8
+	help
+	  Choose for i.MX53
+
 endchoice
 
 config PF_IMX_RAM_PHYS_BASE
@@ -51,3 +58,4 @@ config PF_IMX_RAM_PHYS_BASE
 	default 0x80000000 if PF_IMX_35
 	default 0x90000000 if PF_IMX_51
         default 0x10000000 if PF_IMX_6
+	default 0x70000000 if PF_IMX_53
diff --git kernel/fiasco/src/kern/arm/bsp/imx/Modules kernel/fiasco/src/kern/arm/bsp/imx/Modules
index 5321e62..b43e650 100644
--- kernel/fiasco/src/kern/arm/bsp/imx/Modules
+++ kernel/fiasco/src/kern/arm/bsp/imx/Modules
@@ -10,10 +10,13 @@ PREPROCESS_PARTS	+= $(if $(CONFIG_PF_IMX_51),imx51 imx_epit \
 PREPROCESS_PARTS	+= $(if $(CONFIG_PF_IMX_53),imx53 imx_epit \
                                                     pic_gic pic_gic_mxc_tzic)
 PREPROCESS_PARTS        += $(if $(CONFIG_PF_IMX_6),imx6 pic_gic mptimer)
+
 CONFIG_KERNEL_LOAD_ADDR := $(CONFIG_PF_IMX_RAM_PHYS_BASE)
 
 INTERFACES_KERNEL     += $(if $(CONFIG_PF_IMX_51),gic)
 INTERFACES_KERNEL     += $(if $(CONFIG_PF_IMX_6),gic)
+INTERFACES_KERNEL     += $(if $(CONFIG_PF_IMX_53),gic)
+
 MPCORE_PHYS_BASE      := 0x00a00000
 
 uart_IMPL             += uart-imx
@@ -24,12 +27,14 @@ pic_IMPL              += $(if $(CONFIG_PF_IMX_35),pic-arm-imx)
 pic_IMPL              += $(if $(CONFIG_PF_IMX_51),pic-gic pic-arm-imx51)
 pic_IMPL              += $(if $(CONFIG_PF_IMX_53),pic-gic pic-arm-imx51)
 pic_IMPL              += $(if $(CONFIG_PF_IMX_6),pic-gic pic-arm-imx51)
+
 bootstrap_IMPL        += bootstrap-arm-imx
 timer_IMPL            += $(if $(CONFIG_PF_IMX_21),timer-arm-imx21)
 timer_IMPL            += $(if $(CONFIG_PF_IMX_35),timer-arm-imx_epit)
 timer_IMPL            += $(if $(CONFIG_PF_IMX_51),timer-arm-imx_epit)
 timer_IMPL            += $(if $(CONFIG_PF_IMX_53),timer-arm-imx_epit)
 timer_IMPL            += $(if $(CONFIG_PF_IMX_6),timer-arm-mptimer timer-arm-mptimer-imx6)
+
 timer_tick_IMPL       += timer_tick-single-vector
 kernel_uart_IMPL      += kernel_uart-arm-imx
 reset_IMPL            += reset-arm-imx
diff --git kernel/fiasco/src/kern/arm/bsp/imx/bootstrap-arm-imx.cpp kernel/fiasco/src/kern/arm/bsp/imx/bootstrap-arm-imx.cpp
index 9c37650..1ebb981 100644
--- kernel/fiasco/src/kern/arm/bsp/imx/bootstrap-arm-imx.cpp
+++ kernel/fiasco/src/kern/arm/bsp/imx/bootstrap-arm-imx.cpp
@@ -6,6 +6,7 @@ enum {
 };
 
 //-----------------------------------------------------------------------------
+
 IMPLEMENTATION [arm && imx]:
 void
 map_hw(void *pd)
diff --git kernel/fiasco/src/kern/arm/bsp/imx/config-arm-imx.cpp kernel/fiasco/src/kern/arm/bsp/imx/config-arm-imx.cpp
index 95ebf20..33eb493 100644
--- kernel/fiasco/src/kern/arm/bsp/imx/config-arm-imx.cpp
+++ kernel/fiasco/src/kern/arm/bsp/imx/config-arm-imx.cpp
@@ -14,6 +14,10 @@ INTERFACE [arm && imx53]:
 
 #define TARGET_NAME "i.MX53"
 
+INTERFACE [arm && imx53]:
+
+#define TARGET_NAME "i.MX53"
+
 INTERFACE [arm && imx6]:
 
 #define TARGET_NAME "i.MX6"
diff --git kernel/fiasco/src/kern/arm/bsp/imx/mem_layout-arm-imx.cpp kernel/fiasco/src/kern/arm/bsp/imx/mem_layout-arm-imx.cpp
index 1b849b7..a1b67b8 100644
--- kernel/fiasco/src/kern/arm/bsp/imx/mem_layout-arm-imx.cpp
+++ kernel/fiasco/src/kern/arm/bsp/imx/mem_layout-arm-imx.cpp
@@ -93,6 +93,32 @@ public:
   };
 };
 
+
+INTERFACE [arm && imx && imx53]: // ---------------------------------------
+
+EXTENSION class Mem_layout
+{
+public:
+  enum Virt_layout_imx53 {
+    Timer_map_base       = 0xef1ac000,
+    Uart_map_base        = 0xef1bc000,
+    Watchdog_map_base    = 0xef198000,
+    Gic_cpu_map_base     = 0,
+    Gic_dist_map_base    = 0xef2fc000,
+    Uart_base            = Uart_map_base,
+  };
+
+  enum Phys_layout_imx53 {
+    Device_phys_base_1   = 0x53f00000,
+    Device_phys_base_2   = 0x0ff00000,
+
+    Watchdog_phys_base   = 0x53f98000, // wdog1
+    Timer_phys_base      = 0x53fac000, // epit1
+    Uart_phys_base       = 0x53fbc000, // uart1
+    Gic_dist_phys_base   = 0x0fffc000,
+  };
+};
+
 INTERFACE [arm && imx && imx6]: // -----------------------------------------
 
 EXTENSION class Mem_layout
diff --git l4/mk/platforms/imx53.conf l4/mk/platforms/imx53.conf
new file mode 100644
index 0000000..2f9dfce
--- /dev/null
+++ l4/mk/platforms/imx53.conf
@@ -0,0 +1,4 @@
+PLATFORM_NAME        = "Freescale i.MX53"
+PLATFORM_ARCH        = arm
+PLATFORM_RAM_BASE    = 0x70000000
+PLATFORM_RAM_SIZE_MB = 1024
diff --git l4/pkg/bootstrap/server/src/Make.rules l4/pkg/bootstrap/server/src/Make.rules
index 746e2cf..f2b8b5f 100644
--- l4/pkg/bootstrap/server/src/Make.rules
+++ l4/pkg/bootstrap/server/src/Make.rules
@@ -81,6 +81,7 @@ SUPPORT_CC_arm-imx21       := platform/imx.cc
 SUPPORT_CC_arm-imx35       := platform/imx.cc
 SUPPORT_CC_arm-imx51       := platform/imx.cc
 SUPPORT_CC_arm-imx6        := platform/imx.cc
+SUPPORT_CC_arm-imx53       := platform/imx.cc
 SUPPORT_CC_arm-om          := platform/om.cc
 SUPPORT_CC_arm-kirkwood    := platform/kirkwood.cc
 DEFAULT_RELOC_arm-imx21    := 0x00200000  # because of blob
diff --git l4/pkg/bootstrap/server/src/platform/imx.cc l4/pkg/bootstrap/server/src/platform/imx.cc
index 9da67d7..60edc26 100644
--- l4/pkg/bootstrap/server/src/platform/imx.cc
+++ l4/pkg/bootstrap/server/src/platform/imx.cc
@@ -36,6 +36,9 @@ class Platform_arm_imx : public Platform_single_region_ram
 #elif defined(PLATFORM_TYPE_imx51)
     static L4::Io_register_block_mmio r(0x73fbc000);
     static L4::Uart_imx51 _uart;
+#elif defined(PLATFORM_TYPE_imx53)
+    static L4::Io_register_block_mmio r(0x53fbc000);
+    static L4::Uart_imx51 _uart;
 #elif defined(PLATFORM_TYPE_imx6)
     //static L4::Io_register_block_mmio r(0x02020000); // UART1
     static L4::Io_register_block_mmio r(0x021e8000); // UART2
